### PR TITLE
CarDaoJdbcImpl and DriverDaoJdbcImpl classes iplemented

### PR DESCRIPTION
**Pull request by semyonich**

### CODE REVIEW

**Review by MykolaFatkullin — APPROVED** [review]

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                + "FROM cars c "
+                + "INNER JOIN manufacturers m "
+                + "ON c.manufacturer_id=m.manufacturer_id "
+                + "WHERE c.deleted=false AND c.car_id=?;";
+        String getCarDrivers = "SELECT cd.driver_id, d.driver_name AS driver_name, "
+                + "d.driver_license AS license "
+                + "FROM cars_drivers cd "
+                + "INNER JOIN drivers d "
+                + "ON cd.driver_id=d.driver_id "
+                + "WHERE car_id=? AND d.deleted=false;";
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: let's create method getDrivers(). You can use it here and in getAll() method

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            String manufacturerCountry = resultSet.getString("country");
+            Long manufacturerId = resultSet.getObject("manufacturer_id", Long.class);
+            String carModel = resultSet.getString("model");
+            Long carId = resultSet.getObject("car_id", Long.class);
+            Manufacturer manufacturer = new Manufacturer(manufacturerName, manufacturerCountry);
+            manufacturer.setId(manufacturerId);
+            Car car = new Car(carModel, manufacturer);
+            car.setId(carId);
+            return car;
+        } catch (SQLException e) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: we already catching this exception in the upper method(in each method where you are using makeCar())

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Unable to GET ALL cars from DB", e);
+        }
+        return cars;
+    }
+
+    @Override
+    public Car update(Car car) {
+        String updateCarString = "UPDATE cars SET car_id=?, manufacturer_id=?, car_model=?"
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]:
```suggestion
        String updateCarString = "UPDATE cars SET manufacturer_id=?, car_model=?"
```

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                + " WHERE car_id=? AND deleted=false;";
+        String deleteLinksString = "DELETE FROM cars_drivers WHERE car_id=?;";
+        String newLinksString = "INSERT INTO cars_drivers (car_id, driver_id"
+                + ") VALUES (?,?);";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement carUpdateStatement = connection.prepareStatement(updateCarString);
+                PreparedStatement deleteLinksStatement = connection
+                        .prepareStatement(deleteLinksString);
+                PreparedStatement newLinksStatement = connection.prepareStatement(newLinksString)) {
+            carUpdateStatement.setLong(1, car.getId());
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]:
```suggestion
            carUpdateStatement.setLong(1, car.getId());
```
We are updating by Id. The id by itself won't be changed

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            result = statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Unable to DELETE car from DB, id=" + id, e);
+        }
+        return result > 0;
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        String getCarsByDriverString = "SELECT car_id FROM cars_drivers WHERE driver_id=?;";
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]:
```suggestion
        String getCarsByDriverQuery = "SELECT car_id FROM cars_drivers WHERE driver_id=?;";
```

Comment (dmytro-hryhoruk) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: What if the driver by this id is deleted?

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+    private Driver makeDriver(ResultSet resultSet) {
+        try {
+            String name = resultSet.getString("driver_name");
+            String driverLicense = resultSet.getString("license");
+            Long driverId = resultSet.getObject("driver_id", Long.class);
+            Driver driver = new Driver(name, driverLicense);
+            driver.setId(driverId);
+            return driver;
+        } catch (SQLException e) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java:192]: same

File: src/main/java/core/basesyntax/dao/jdbc/DriverDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            }
+            return driver;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Unable to CREATE driver in DB " + driver, e);
+        }
+    }
+
+    @Override
+    public Optional<Driver> get(Long id) {
+        String getString = "SELECT * FROM drivers WHERE driver_id=? AND deleted=false;";
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/DriverDaoJdbcImpl.java]:
```suggestion
        String getDriverQuery = "SELECT * FROM drivers WHERE driver_id=? AND deleted=false;";
```

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java (hunk 1 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: done

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java (hunk 2 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: done

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java (hunk 4 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: done

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java (hunk 3 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: done

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java (hunk 5 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java]: fixed

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java (hunk 6 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java:192]: removed

File: src/main/java/core/basesyntax/dao/jdbc/DriverDaoJdbcImpl.java (hunk 7 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/DriverDaoJdbcImpl.java]: renamed everywhere

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        Car car = null;
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement carStatement = connection.prepareStatement(getCarQuery)) {
+            carStatement.setLong(1, id);
+            ResultSet carResultSet = carStatement.executeQuery();
+            while (carResultSet.next()) {
+                car = makeCar(carResultSet);
+            }
+            List<Driver> thiCarDrivers = getCarDrivers(id);
+            car.setDrivers(thiCarDrivers);
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java:58]: let's fetch and set drivers only if carResultSet.next() == true. Also let's create Car link inside try block and add return statement here too

**Review by dmytro-hryhoruk — APPROVED** [review]

File: src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java (hunk 8 above)
Comment (semyonich) [reply src/main/java/core/basesyntax/dao/jdbc/CarDaoJdbcImpl.java:58]: I will add this to the next homework, to simplify git merges

**Pull request merged**